Chronopost carrier integration

Ftayri opened the pull request:
karrio.chronopost

danh91 commented:
Hi @Ftayri,

I took a deeper look at what you have here and made some improvements to simplify the integration.
You are on the right track.

I will leave a couple of comments around and we can schedule a call to walk through edge cases.

danh91 commented on schemas/chronopost/generate.sh:
```diff
@@ -0,0 +1,7 @@
+find "${LIB_MODULES}" -name "*.py" -exec rm -r {} \;
+touch "${LIB_MODULES}/__init__.py"
+
+generateDS --no-namespace-defs -o "${LIB_MODULES}/services.py" "${SCHEMAS}/ShippingServiceWS.xml"
```
I looked at the documentation and for the full integration of the shipping service, you will also need the QuickCost schema to parse the rating responses

https://ws.chronopost.fr/quickcost-cxf/QuickcostServiceWS?wsdl

danh91 replied:
Actually, you might as well generate the schema types for TrackingService since the response for cancel shipment (cancelSkyBill) response schema will be required

https://ws.chronopost.fr/tracking-cxf/TrackingServiceWS?wsdl

danh91 commented on sdk/extensions/chronopost/karrio/mappers/chronopost/proxy.py:
```diff
@@ -0,0 +1,45 @@
+        )
+
+    def get_rates(self, request: Serializable) -> Deserializable:
+        return super().get_rates(request)
```
You will need to implement the API request here as per the QuickCost service documentation

```
https://ws.chronopost.fr/quickcost-cxf/QuickcostServiceWS/quickCost?
accountNumber=[accountNumber]&password=[password]&depCode=28500&arrCode=75001&weight=1&produ
ctCode=1&type=M
```

danh91 commented on sdk/extensions/chronopost/karrio/mappers/chronopost/proxy.py:
```diff
@@ -0,0 +1,45 @@
+    def create_shipment(self, request: Serializable) -> Deserializable[str]:
+        response = self._send_request(
+            request,
+            soapaction=f"{self.settings.server_url}#createShipment",
```
I don't think the `soapaction` will be necessary for Chronopost in general

danh91 replied:
```suggestion
```

danh91 commented on sdk/extensions/chronopost/karrio/mappers/chronopost/proxy.py:
```diff
@@ -0,0 +1,45 @@
+    def cancel_shipment(self, request: Serializable) -> Deserializable[str]:
+        response = self._send_request(
+            request,
+            soapaction=f"{self.settings.server_url}#deleteShipment",
```
```suggestion
```

danh91 commented on sdk/extensions/chronopost/karrio/providers/chronopost/shipment/cancel.py:
```diff
@@ -0,0 +1,54 @@
+        lambda request: settings.serialize(
+            request, "createShipment", settings.server_url
+        ),
+        logged=True,
```
```suggestion
```

This is won't be necessary anymore

danh91 commented on sdk/extensions/chronopost/karrio/providers/chronopost/shipment/create.py:
```diff
@@ -0,0 +1,172 @@
+                shipperName=shipper.company_name,
+                shipperName2=shipper.company_name,
+                shipperPreAlert=0,
+                shipperCivility="M",
```
Is this required or we can keep it `None`

Ftayri replied:
Apparently it is required, the service would fail if I don't include it in the envelope.

danh91 replied:
No problem. What we can do is we can declare an option for that so it can be optionally customized but we will keep a default value as you did.

By any chance, is there another option that is more neutral instead of `M`?

danh91 commented on sdk/extensions/chronopost/karrio/providers/chronopost/units.py:
```diff
@@ -0,0 +1,39 @@
+    chronopost_classic_international = "44"
+
+
+class Option(Flag):
+    chronopost_delivery_on_monday = Spec.asKey("1")
+    chronopost_delivery_on_saturday = Spec.asKey("6")
+    chronopost_delivery_normal = Spec.asKey("0")
```
I simplified the definition of shipping options.
Please merge main to use the new approach

danh91 reviewed (changes requested):
It's looking good overall, I will recommend you merge the latest `main` and leverage the new `cli` and helpers to move faster.

I am updating the documentation now to add the new `cli` and the improvements.

Ftayri commented:
Hi @danh91 ,
I updated the code to match the new structure. I also added the new schemas to implement cancelling shipments and fetching rates. For fetching rates I decided to use `calculateProdcuts` from the QuickCost webservice to fetch the rates. `quickCost` needs the service code in the request in order to fetch its rate and the rate of other services for some reason.
I'm also not sure if parsing responses and extracting details in both create.py and rate.py are well implemented.

danh91 commented:
No problem. I will take a look shortly and give you some guidance for the next steps.

danh91 commented on .vscode/settings.json:
```diff
@@ -12,4 +12,7 @@
     "**/templates/**/*": "django-txt",
     "**/requirements{/**,*}.{txt,in}": "pip-requirements"
   },
+  "python.analysis.extraPaths": [
+    "./schemas/chronopost"
+  ],
```
Instead of this. Please add the package to the dev dependency

`requirements.sdk.dev.txt`
```txt
# Schemas update
-e ./schemas/chronopost
```
same for `requirements.server.dev.txt`
```txt
# Schemas update
-e ./schemas/chronopost
```

danh91 commented on schemas/chronopost/schemas/QuickcostServiceWS.xml:
```diff
@@ -0,0 +1,121 @@
+<xs:schema xmlns:tns="http://cxf.quickcost.soap.chronopost.fr/" xmlns:xs="http://www.w3.org/2001/XMLSchema" attributeFormDefault="unqualified" elementFormDefault="unqualified" targetNamespace="http://cxf.quickcost.soap.chronopost.fr/">
```
Please rename this files into `.xsd` instead of `.xml`

danh91 commented on schemas/chronopost/schemas/ShippingServiceWS.xml:
```diff
@@ -0,0 +1,1542 @@
+<xs:schema xmlns:tns="http://cxf.shipping.soap.chronopost.fr/" xmlns:xs="http://www.w3.org/2001/XMLSchema" attributeFormDefault="unqualified" elementFormDefault="unqualified" targetNamespace="http://cxf.shipping.soap.chronopost.fr/">
```
Please rename this files into `.xsd` instead of `.xml`

danh91 commented on schemas/chronopost/schemas/TrackingServiceWS.xml:
```diff
@@ -0,0 +1,265 @@
+<xs:schema xmlns:tns="http://cxf.tracking.soap.chronopost.fr/" xmlns:xs="http://www.w3.org/2001/XMLSchema" attributeFormDefault="unqualified" elementFormDefault="unqualified" targetNamespace="http://cxf.tracking.soap.chronopost.fr/">
```
Please rename this files into `.xsd` instead of `.xml`

danh91 commented on sdk/extensions/chronopost/karrio/mappers/chronopost/mapper.py:
```diff
@@ -0,0 +1,46 @@
+    TrackingDetails,
+    Message,
+)
+from karrio.universal.providers.rating import parse_rate_response, rate_request
```
```suggestion
```

We will need to implement the chronopost-specific rate request and response so we can remove this import.

danh91 commented on sdk/extensions/chronopost/karrio/providers/chronopost/units.py:
```diff
@@ -0,0 +1,57 @@
+    chronopost_classic_international = "44"
+
+
+class ShippingOption(lib.Flag):
+    chronopost_delivery_on_monday = OptionEnum("1")
+    chronopost_delivery_on_saturday = OptionEnum("6")
+    chronopost_delivery_normal = OptionEnum("0")
```
That's good

danh91 reviewed (changes requested):
Overall, You've made some good progress.
schedule a time here for us to do some pair programming and wrap it up
https://calendly.com/karrioapi/chat

Ftayri commented:
@danh91 I managed to run the shipment integration tests, fixed a bunch of issues but I couldn't solve 2 problems. The first one is that the label isn't parsed for some reason. As for the second problem, cancelling shipment returns an error with code 4 saying document is empty.
Both problematic responses are `ParsedShipmentResponse` and `ParsedCancelShipmentResponse` in shipment.py if you want to look at them.

danh91 commented on sdk/extensions/chronopost/karrio/providers/chronopost/shipment/create.py:
```diff
@@ -0,0 +1,204 @@
+    response: lib.Element, settings: provider_utils.Settings
+) -> typing.Tuple[models.ShipmentDetails, typing.List[models.Message]]:
+    errors = provider_error.parse_error_response(response, settings)
+    shipment = (
+        _extract_details(response, settings)
+        if lib.find_element("resultMultiParcelValue", response, first=True) is not None
+        else None
+    )
```
```suggestion
    shipment_node = lib.find_element("resultMultiParcelValue", response, first=True)
    shipment = (
        _extract_details(result, settings)
        if shipment_node is not None else None
    )
```

I think that is the reason why the shipment parsing is failing.

Ftayri replied:
I also changed `shipment = lib.to_object(resultMultiParcelValue, response)` to ` shipment = lib.find_element("resultMultiParcelValue", response, resultMultiParcelValue, first=True)` in `_extract_details()` and that fixed it.

danh91 replied:
🤔 I made a mistake actually. there was no `result` variable it was supposed to be `shipment_nodes`.

Once that is done, I don't think you need to do this change: 

> I also changed `shipment = lib.to_object(resultMultiParcelValue, response)` to ` shipment = lib.find_element("resultMultiParcelValue", response, resultMultiParcelValue, first=True)` in `_extract_details()` and that fixed it.

danh91 reviewed (changes requested):
@Ftayri check these suggestions for the issue with the shipment response parsing and the cancel request

Ftayri commented:
With that, all the current tests are validated! I'll see if I can get a successful shipment cancel response and add it to the tests. I'll also double check the products and services of Chronopost to make sure quickCost is the right request for fetching rates.

danh91 commented on sdk/extensions/chronopost/tests/chronopost/shipment.py:
```diff
@@ -0,0 +1,255 @@
+        "meta": {},
+    },
+    [],
+]
```
This is still incorrect

danh91 replied:
My bad 👍🏿

danh91 commented on sdk/extensions/chronopost/tests/chronopost/shipment.py:
```diff
@@ -0,0 +1,255 @@
+            "message": "Document is empty, line 1, column 1 (<string>, line 1)",
+        }
+    ],
+]
```
And that one too.

Ftayri replied:
Can you verify the commits again? I think that you commented on an outdated commit 🤔

danh91 replied:
my bad. Thanks

danh91 commented on sdk/extensions/chronopost/tests/chronopost/shipment.py:
```diff
@@ -0,0 +1,253 @@
+    "shipment_identifier": "XP696982485FR",
+}
+
+ParsedShipmentResponse = [
+    {
+        "carrier_id": "chronopost",
+        "carrier_name": "chronopost",
+        "docs": {},
+        "meta": {},
+        "shipment_identifier": "XP696982485FR",
+        "tracking_number": "XP696982485FR",
+    },
+    [],
+]
```
actually, the label is still missing

Ftayri replied:
For some reason the parsed response tag for the label was `pdfetiquette` instead of `pdfEtiquette` 😆 , now there is a problem with serializing the shipment details. I'm getting a circular reference detected that's definitely coming from the label. This is what the parsed response looks like before serializing it. `(ShipmentDetails(carrier_name='chronopost', carrier_id='chronopost', tracking_number='XP696982485FR', shipment_identifier='XP696982485FR', docs=Documents(label=b'\x89PNG\r\n\x1a\n\x00\x00\x00\rIHDR\x00\x00\x00\x01\x00\x00\x00\x01\x08\x02\x00\x00\x00\x90wS\xde\x00\x00\x00\x0cIDATx^c\x98\xc9\xf0\x1f\x00\x02\xce\x01\x99[H\x07\x99\x00\x00\x00\x00IEND\xaeB`\x82', invoice=None), selected_rate=None, label_type=None, meta={}, id=None), [])`

danh91 replied:
I think it is returning a base64 bytes. So you need to encode it to sting

```
label = base64.b64encode(shipment.pdfetiquette).decode("utf-8")
```

Ftayri replied:
I thought so too, so labels from other carriers are always string? 🤔

danh91 replied:
They all come as a string but the XML parser will sometime convert it into bytes.

Ftayri commented:
I update rating request to use `calculateProducts`, `quickCost` actually returns a bunch of shipping options, to supplement the service, that are not used anymore. I also started implementing tracking. I'll have to find a working tracking response to be able to implement it correctly, the test account I'm using can't do that.

Ftayri commented:
@danh91 I'm having trouble with parsing tracking response, the schema is weird and I was relying on an actual envelope response to implement the parsing but it's not working, I'm only getting the carrier settings in the parsed response and none of the tracking details `[[], [{'carrier_id': 'chronopost', 'carrier_name': 'chronopost'}]]`

Ftayri commented:
I managed to parse all of the details except the skybillnumber 😅

Ftayri commented:
Okay I managed to fix the issue, though I'm not sure if my implementation is clean so I'd appreciate your feedback!

danh91 commented on sdk/extensions/chronopost/karrio/mappers/chronopost/proxy.py:
```diff
@@ -0,0 +1,55 @@
+import urllib.parse
+import karrio.lib as lib
+import karrio.api.proxy as proxy
+from karrio.core.utils import (
+    Envelope,
+    XP,
+    request as http,
+    exec_parrallel,
+)
```
```suggestion
```

I would suggest using `lib` instead because it has all these helpers and it is better typed.

danh91 commented on sdk/extensions/chronopost/karrio/mappers/chronopost/proxy.py:
```diff
@@ -0,0 +1,55 @@
+
+        response: typing.List[str] = exec_parrallel(get_tracking, request.serialize())
+
+        return lib.Deserializable(XP.bundle_xml(xml_strings=response), XP.to_xml)
```
```suggestion
        return lib.Deserializable(response, lib.to_element)
```
You can simplify it like this because `lib.to_element` function will bundle the XML responses under one element `<bundle>....</bundle>`.

danh91 commented on sdk/extensions/chronopost/karrio/providers/chronopost/tracking.py:
```diff
@@ -0,0 +1,84 @@
+
+
+def parse_tracking_response(
+    responses: typing.List[lib.Element],
```
since responses have been bundled together, you will get a single parent element for all

```suggestion
    responses: lib.Element,
```

danh91 commented on sdk/extensions/chronopost/karrio/providers/chronopost/tracking.py:
```diff
@@ -0,0 +1,84 @@
+    responses: typing.List[lib.Element],
+    settings: provider_utils.Settings,
+) -> typing.Tuple[typing.List[models.TrackingDetails], typing.List[models.Message]]:
+    tracking_nodes: typing.List[chronopost.eventInfoComp] = [
+        lib.find_element("events", response, chronopost.eventInfoComp)
+        for response in responses
+    ]
```
I would suggest getting the `listEventInfoComp` node instead because it is specific to one `skybillnumber` and gives you the events too.
```suggestion
    tracking_nodes: typing.List[chronopost.eventInfoComp] = [
        lib.find_element("listEventInfoComp", response, chronopost.eventInfoComp)
        for response in responses
    ]
```

Ftayri replied:
I was going with this approach at first, but when I tried to get an attribute like eventDate from listEventInfoComps.events, nothing shows up in the editor. It lured me into thinking I was using the wrong classes 😅 but apparently it works.

danh91 commented on sdk/extensions/chronopost/karrio/providers/chronopost/tracking.py:
```diff
@@ -0,0 +1,84 @@
+        lib.find_element("events", response, chronopost.eventInfoComp)
+        for response in responses
+    ]
+    errors = [
+        provider_error.parse_error_response(result, settings) for result in responses
+    ]
```
You don't need the wrapping list. since it will the parse error function will retrieve the list of errors
```suggestion
    errors = provider_error.parse_error_response(result, settings)
```

danh91 commented on sdk/extensions/chronopost/karrio/providers/chronopost/tracking.py:
```diff
@@ -0,0 +1,84 @@
+    errors = [
+        provider_error.parse_error_response(result, settings) for result in responses
+    ]
+    tracking_details = [
+        _extract_details(tracking_nodes[t], settings, responses[t])
+        for t in range(len(tracking_nodes))
+    ]
```
Once you fix the selection above to get the list of `listEventInfoComp` instead, you can just pass that here
```suggestion
    tracking_details = [
        _extract_details(node, settings) for node in tracking_nodes
    ]
```

danh91 commented on sdk/extensions/chronopost/karrio/providers/chronopost/tracking.py:
```diff
@@ -0,0 +1,84 @@
+    return tracking_details, errors
+
+
+def _extract_details(
+    detail: typing.List[chronopost.eventInfoComp],
+    settings: provider_utils.Settings,
+    response: typing.List[lib.Element],
+) -> models.TrackingDetails:
```
With the fixes above, all you will need is the node and settings as parameters
```suggestion
def _extract_details(
    detail: chronopost.eventInfoComp,
    settings: provider_utils.Settings,
) -> models.TrackingDetails:
```

danh91 commented on sdk/extensions/chronopost/karrio/providers/chronopost/tracking.py:
```diff
@@ -0,0 +1,84 @@
+    settings: provider_utils.Settings,
+    response: typing.List[lib.Element],
+) -> models.TrackingDetails:
+    tracking_number: chronopost.listEventInfoComps = lib.find_element(
+        "listEventInfoComp", response, chronopost.listEventInfoComps, first=True
+    )
```
This won't be necessary anymore.

danh91 replied:
This won't be necessary anymore
```suggestion

```

danh91 commented on sdk/extensions/chronopost/karrio/providers/chronopost/tracking.py:
```diff
@@ -0,0 +1,84 @@
+    tracking_number: chronopost.listEventInfoComps = lib.find_element(
+        "listEventInfoComp", response, chronopost.listEventInfoComps, first=True
+    )
+    return models.TrackingDetails(
+        carrier_id=settings.carrier_id,
+        carrier_name=settings.carrier_name,
+        tracking_number=tracking_number.skybillNumber,
+        events=[
+            models.TrackingEvent(
+                date=lib.fdate(event.eventDate, "%Y-%m%dT%H%:%M:%S"),
+                description=event.eventLabel,
+                code=event.code,
+                time=lib.ftime(event.eventDate, "%Y-%m%dT%H%:%M:%S"),
+                location=event.officeLabel,
+            )
+            for event in detail
+        ],
+        estimated_delivery=None,
+        delivered=True if "D" in [event.code for event in detail] else False,
+    )
```
Now, you should be able to do
```suggestion
    return models.TrackingDetails(
        carrier_id=settings.carrier_id,
        carrier_name=settings.carrier_name,
        tracking_number=detail.skybillNumber,
        events=[
            models.TrackingEvent(
                date=lib.fdate(event.eventDate, "%Y-%m%dT%H%:%M:%S"),
                description=event.eventLabel,
                code=event.code,
                time=lib.ftime(event.eventDate, "%Y-%m%dT%H%:%M:%S"),
                location=event.officeLabel,
            )
            for event in detail.events
        ],
        delivered=any([event.code == "D" for event in detail.events]),
    )
```

danh91 commented on sdk/extensions/chronopost/tests/chronopost/tracking.py:
```diff
@@ -0,0 +1,267 @@
+            "tracking_number": "89108749065090",
+        }
+    ],
+    [[]],
```
This should just be 
```suggestion
    [],
```

Fixing the error parsing above should correct that

danh91 commented on sdk/extensions/chronopost/tests/chronopost/tracking.py:
```diff
@@ -0,0 +1,267 @@
+            print(lib.to_dict(parsed_response))
+            self.assertListEqual(lib.to_dict(parsed_response), ParsedTrackingResponse)
+
+    # def test_parse_error_response(self):
+    #     with patch("karrio.mappers.chronopost.proxy.lib.request") as mock:
+    #         mock.return_value = ErrorResponse
+    #         parsed_response = (
+    #             karrio.Tracking.fetch(self.TrackingRequest).from_(gateway).parse()
+    #         )
+
+    #         self.assertListEqual(lib.to_dict(parsed_response), ParsedErrorResponse)
```
I would recommend adding a the test for tracking error parsing.

Ftayri replied:
Okay, I wanted to leave parsing errors for last because it's kind of weird. All chronopost responses are in a `<return></return>`  tag and in that tag you always find `errorCode` which is 0 if everything is fine and `errorMessage` that is sent when an error occures. I couldn't figure out how to parse this.

danh91 commented on sdk/extensions/chronopost/karrio/mappers/chronopost/proxy.py:
```diff
@@ -0,0 +1,55 @@
+                request=lib.Serializable(track_request),
+            )
+
+        response: typing.List[str] = exec_parrallel(get_tracking, request.serialize())
```
You can replace `exec_parrallel` by
```suggestion
        response: typing.List[str] = lib.run_concurently(get_tracking, request.serialize())
```

danh91 reviewed (changes requested):
Great work. You are almost there.
I left a couple of comment for the tracking integration

danh91 commented on sdk/extensions/chronopost/karrio/providers/chronopost/tracking.py:
```diff
@@ -0,0 +1,84 @@
+) -> lib.Serializable:
+    request = [
+        lib.Envelope(
+            Header=lib.Header(),
```
Is the Header required?

Ftayri commented:
I changed what's needed in tracking and changed `parse_error_response` in provider error to be able to correctly parse the errors. thought I'm not sure if it's the optimal way to do it 🤔 I also added error responses for all of the tests and validated them. As always I would greatly appreciate your input!

danh91 commented on sdk/extensions/chronopost/karrio/providers/chronopost/error.py:
```diff
@@ -0,0 +1,30 @@
+import typing
+from karrio.core.models import Message
+from karrio.providers.chronopost.utils import Settings
+from chronopost_lib import quickcostservice, shippingservice, trackingservice
+import karrio.lib as lib
+
+
+def parse_error_response(
+    response: lib.Element, settings: Settings, response_type: str
+) -> typing.List[Message]:
+    element = (
+        shippingservice.shippingMultiParcelV5Response
+        if response_type == "shippingMultiParcelV5Response"
+        else trackingservice.cancelSkybillResponse
+        if response_type == "cancelSkybillResponse"
+        else quickcostservice.calculateProductsResponse
+        if response_type == "calculateProductsResponse"
+        else trackingservice.trackSkybillV2Response
+    )
+    errors = lib.find_element(response_type, response, element)
+    return [
+        Message(
+            carrier_id=settings.carrier_id,
+            carrier_name=settings.carrier_name,
+            message=error.return_.errorMessage,
+            code=error.return_.errorCode,
+        )
+        for error in errors
+        if error.return_.errorCode != 0
+    ]
```
This works. Here is a proposition. to make it lighter,
The good thing about the XML parsing library is that if it is fail-safe and will ignore nodes it doesn't know and process the one it recognizes.

So you don't need to know if it is a `shippingMultiParcelV5Response` or `calculateProductsResponse`. 
Either class will be able to parse all the responses and collect the errorCode and errorMessage since those are the fields they have in common.

Therefore, we just need to take one class that knows how to parse a `<return>...</return>` node and used it on all the responses to parse and check if they have an `errorCode` different from 0.

Please try this and let me know.

```suggestion
import typing
from karrio.core.models import Message
from karrio.providers.chronopost.utils import Settings
import chronopost_lib.shippingservice as ups
import karrio.lib as lib

ReturnType = ups. resultMultiParcelExpeditionValue

def parse_error_response(response: lib.Element, settings: Settings) -> typing.List[Message]:
    errors = lib.find_element("return", response, ReturnType)
    
    return [
        Message(
            carrier_id=settings.carrier_id,
            carrier_name=settings.carrier_name,
            message=error.errorMessage,
            code=error.errorCode,
        )
        for error in errors
        if error.errorCode != 0
    ]
```

Ftayri replied:
Mhm, this is definitely better!

danh91 commented:
Overall, fantastic job looks pretty much ready to merge.

Ftayri commented:
@danh91 just waiting for your approval 😃

danh91 commented:
Hi @Ftayri,

Few final things for the merge.
- please run `black` on `sdk/extensions/chronopost` (just the extension not the schema)
- rebase your branch on the `upstream karrio-2022.8` release branch
- Then let me know and I will merge

Very excited. Thank you

Ftayri commented:
Just to confirm, I should `git fetch upstream` then `git checkout karrio-2022.8` and `git rebase upstream/karrio-2022.8` right? 😅 I'm worrying too much.

danh91 commented:
> Just to confirm, I should `git fetch upstream` then `git checkout karrio-2022.8` and `git rebase upstream/karrio-2022.8` right? 😅 I'm worrying too much.

Actually, just run black. I will take care of the rebase.

Ftayri commented:
Already done.
